### DISCUSSION TITLE
Add filtering options for The Main/ComboViewer using the code of the FX swatch viewer

### DISCUSSION THREAD

**Discussion by oliver-dreamwalker** [opentoonz/opentoonz#6571] [Ideas]
### Is your feature request related to a problem? Please describe

The lack of a **Filtering**(I guess OT uses Nearest-neighbor) in the **Main viewer/Combo viewer** is not ideal for sketching or previewing the final render of the animation, since lines disappear and/or they don't look very smooth when **zoomed out**.

However, for some reason, the Swatch Viewer has some kind of scale filtering
and it's a better preview/representation of the final render.

Here's the difference between the **Main Viewer** and the **Swatch Viewer**:
 ![better filtering](https://user-images.githubusercontent.com/66680990/206798203-756049a6-65cc-4e80-a84b-35fe8329d8fe.PNG)

**Bilinear Filtering** = **smoother** looking lines.

There's a reason why many art programs has Bilinear Filtering by default.
 

### Describe the solution you'd like

I think the solution would be to recycle some of the code used for the Swatch Viewer and use it on the **Main Viewer**. 
In case someone needs the default Nearest-neighbor, add the options for the user to select.  
For example:
![filter](https://user-images.githubusercontent.com/66680990/206800303-34e34859-d064-409b-88f7-c85239a8aa41.PNG)

This feature might fix:
Issue #4136
Issue #3441

**Comment by artisteacher** [comment]
That's quite a difference! I wonder if there's a performance hit with the larger window?

**Comment by oliver-dreamwalker** [comment]
> That's quite a difference! I wonder if there's a performance hit with the larger window?

@artisteacher,  I kind of wondered that too, but I when I resized both windows to have the same size, the **Swatch Viewer** (the better looking one) rendered it slightly faster. Maybe @shun-iwasawa has the definitive answer.

**Comment by shun-iwasawa** [comment]
The swatch viewer renders the current viewer size as the camera size. Since the size of the swatch viewer is usually small, the calculation is fast. And since the viewer does not resample the rendered image when it is displayed, it has a smooth appearance as you pointed.
Normal previews are rendered with the camera size set in the output settings, and resamples it with nearest neighbor when displayed in the viewer's size.
I believe that resampling with nearest neighbor is mainly for performance reasons, so it might be a good idea to allow users with a fast-enough working environment to choose resampling with Bilinear or other resampling options.

**Comment by oliver-dreamwalker** [comment]
@shun-iwasawa, **just curious**, is it hard to add resampling options to the viewer ?

**Comment by flurick** [comment]
In a quick test there is a bilinear option ready but it hinted at being a "fast" version, I assume that means lower quality. I guess there are more steps/flags needed because both setting the opengl option to a different mode and using toonz bilinear over nearest neighbor showed no difference in a first try at this.

I also found this that may or may not be ralated: https://github.com/opentoonz/opentoonz/blob/d2c95a8ff8b725a19501dca0017cc9e011776fc5/toonz/sources/toonzlib/stagevisitor.cpp#L533

**Comment by oliver-dreamwalker** [comment]
I heard bilinear is already one of the fastest types of filtering, I don't know what a "fast"/lower quality version of it would even look like.

**Comment by RodneyBaker** [comment]
@oliver-dreamwalker You may want to test the latest nightly release as it includes some features designed to optimize previewing of the viewer.

**Comment by oliver-dreamwalker** [comment]
> @oliver-dreamwalker You may want to test the latest nightly release as it includes some features designed to optimize previewing of the viewer.

I tested it, but still does not have bilinear filtering in the viewer.

**Comment by RodneyBaker** [comment]
Interesting.
I wonder what this setting is for then:

![image](https://user-images.githubusercontent.com/2576432/217362957-48a22b4b-3597-4e64-b194-4b0eb7b51ca5.png)

Perhaps that is for when the Preview is rendererd to disk. (Not sure)

**Comment by RodneyBaker** [comment]
I'm definitlely seeing a big difference.  I do need to Regenerate Preview to get different Resample Balance to take effect though.
It can help to use the subcamera to see the difference from inside the subcamera versus outside.

![image](https://user-images.githubusercontent.com/2576432/217398626-eb23faac-b55a-4c5e-80b9-a167f0427718.png)

**Comment by oliver-dreamwalker** [comment]
The Resample Balance WORKS, the problem the viewer still has a Nearest Neighbor filtering on top of that (It's like watching 4K movie on a SD TV). Somehow, the Swatch Viewer has some kind of filtering that it's NOT Nearest Neighbor and it is a good example of what I would like for the main viewer too.

**Comment by flurick** [comment]
Apparently the swatch viewer does Not use bilinear
https://github.com/opentoonz/opentoonz/blob/34caff449113c2ffbeead362b044e557bb600791/toonz/sources/toonzqt/swatchviewer.cpp#L573
Auto translated: 
// it would be tempting to use the TRop::Bilinear filter (faster), // but the quality suffers a lot

**Comment by oliver-dreamwalker** [comment]
It was a guess since bilinear is very common but I don't think the swatch view uses nearest neighbor either.
Do you know what kind of filter the Swatch Viewer uses ?
How viable it would be to copy and adapt the code to work with the other viewer windows ?

**Comment by ChengduLittleA** [comment]
This seems to be related to the implementation of `quickPut()`. Not sure if there's a very straightforward approach, but I think I can hack it to get it to work.

**Comment by oliver-dreamwalker** [comment]
Interesting...

I wonder if that will work.

**Comment by ChengduLittleA** [comment]
Due to it's using color ID, bilinear filtering might be a nasty implementation under current structure of `quickPut()`, I think I might as well just draw the image 4 times in the viewport to mitigate the issue.

**Comment by oliver-dreamwalker** [comment]
@ChengduLittleA, what kind of filtering does the Swatch Viewer uses? I heard it is **NOT** bilinear.

**Comment by ChengduLittleA** [comment]
@oliver-dreamwalker I haven't even found out where the swatch viewer is... But whatever it uses it does have filtering.

The `quickput()`, `onRasterImage()` and the overall design of `TRop` is so messy it's really hard to add anything to it myself...

**Comment by flurick** [comment]
> @oliver-dreamwalker I haven't even found out where the swatch viewer is... But whatever it uses it does have filtering.
> 
> The `quickput()`, `onRasterImage()` and the overall design of `TRop` is so messy it's really hard to add anything to it myself...

Yeah it's a lot of details. Could it be that the [quickPuts() ](https://github.com/opentoonz/opentoonz/blob/master/toonz/sources/common/trop/tover.cpp#L490) in [stagevisitor.cpp RasterPainter::flushRasterImages() ](https://github.com/opentoonz/opentoonz/blob/d8192195f4fc53729290a8135ed9fba8b3a5fa90/toonz/sources/toonzlib/stagevisitor.cpp#L459) can just be replace with one of the [over()](https://github.com/opentoonz/opentoonz/blob/d8192195f4fc53729290a8135ed9fba8b3a5fa90/toonz/sources/common/trop/tover.cpp#L441) functions that match?

Edit: Just scrolling around a bit and yes, this is what swatchviewer.cpp does! But how is the performance?

**Comment by ChengduLittleA** [comment]
@flurick not sure, The last time I looked into it, it seems that the data format doesn't match. `over` doesn't seem to have matching format, but I could be wrong. The versioning of the function is all over the place lol

**Comment by flurick** [comment]
@ChengduLittleA oh, maybe its this then https://github.com/opentoonz/opentoonz/blob/d8192195f4fc53729290a8135ed9fba8b3a5fa90/toonz/sources/common/trop/quickput.cpp#L5188
(Only do filtering if both layers are TRaster32P)

**Comment by flurick** [comment]
Got nothing on this one still.
Only that there is an option to [opengl texture](https://github.com/opentoonz/opentoonz/blob/8ce93e4e810d4ee40c6c08d486dbac354a72a6e5/toonz/sources/include/toonz/imagepainter.h#LL64C15) instead?

**Comment by oliver-dreamwalker** [comment]
@flurick, maybe @shun-iwasawa can help you with that.

**Comment by ChengduLittleA** [comment]
> (Only do filtering if both layers are TRaster32P)

@flurick Yeah that's the problem.

Using texture could be ok for raster layers but not for those color look up layers and vector.
